main/lsp/ErrorReporter.cc

```diff
@@ -0,0 +1,104 @@
+#include "core/lsp/TypecheckEpochManager.h"
+#include "main/lsp/LSPMessage.h"
+#include "main/lsp/LSPOutput.h"
+#include "main/lsp/json_types.h"
+
+namespace sorbet::realmain::lsp {
+using namespace std;
+ErrorReporter::ErrorReporter(shared_ptr<const LSPConfiguration> config) : config(move(config)) {}
+
+// Used for unit tests
+const std::vector<ErrorStatus> &ErrorReporter::getFileErrorStatuses() const {
+    return fileErrorStatuses;
+};
+
+vector<core::FileRef> ErrorReporter::filesUpdatedSince(u4 epoch) {
+    vector<core::FileRef> filesUpdatedSince;
+    for (size_t i = 1; i < fileErrorStatuses.size(); ++i) {
+        ErrorStatus fileErrorStatus = fileErrorStatuses[i];
+        if (fileErrorStatus.lastReportedEpoch >= epoch && fileErrorStatus.hasErrors) {
+            filesUpdatedSince.push_back(core::FileRef(i));
+        }
+    }
+    return filesUpdatedSince;
+}
+
+void ErrorReporter::pushDiagnostics(u4 epoch, core::FileRef file, const vector<unique_ptr<core::Error>> &errors,
+                                    const core::GlobalState &gs) {
+    ENFORCE(file.exists());
+    if (file.data(gs).epoch > epoch) {
+        return;
+    }
+
+    ErrorStatus &fileErrorStatus = getFileErrorStatus(file);
+    if (fileErrorStatus.lastReportedEpoch > epoch) {
+        return;
+    }
+
+    // Update epoch of any file that is typechecked
+    if (fileErrorStatus.lastReportedEpoch < epoch) {
+        fileErrorStatus.lastReportedEpoch = epoch;
+    }
+
+    // If errors is empty and the file had no errors previously, break
+    if (errors.empty() && fileErrorStatus.hasErrors == false) {
+        return;
+    }
+
+    fileErrorStatus.hasErrors = !errors.empty();
+
+    const string uri = config->fileRef2Uri(gs, file);
+    config->logger->debug("[ErrorReporter] Sending diagnostics for file {}, epoch {}", uri, epoch);
+    vector<unique_ptr<Diagnostic>> diagnostics;
+    for (auto &error : errors) {
+        ENFORCE(!error->isSilenced);
+
+        auto range = Range::fromLoc(gs, error->loc);
+        if (range == nullptr) {
+            continue;
+        }
+
+        auto diagnostic = make_unique<Diagnostic>(std::move(range), error->header);
+        diagnostic->code = error->what.code;
+        diagnostic->severity = DiagnosticSeverity::Error;
+
+        vector<unique_ptr<DiagnosticRelatedInformation>> relatedInformation;
+        for (auto &section : error->sections) {
+            string sectionHeader = section.header;
+
+            for (auto &errorLine : section.messages) {
+                string message = errorLine.formattedMessage.length() > 0 ? errorLine.formattedMessage : sectionHeader;
+                auto location = config->loc2Location(gs, errorLine.loc);
+                if (location == nullptr) {
+                    continue;
+                }
+                relatedInformation.push_back(make_unique<DiagnosticRelatedInformation>(std::move(location), message));
+            }
+            // Add link to error documentation.
```

Comment:
Ah, I think this has to go outside of this `error->sections` loop. It should always be the last `relatedInformation` we push -- it provides a single link for more information on the error.

I think this is why the recorded tests were failing.